### PR TITLE
Fix bug in GrpcBrokerRequestHandler: exceptions are suppressed

### PR DESCRIPTION
**Pull request by dongxiaoman**
## Description

In our `GrpcBrokerRequestHandler` if there is an exception in gRPC calls, the exceptions are suppressed and there is no trace or error handling. The handler will consider all calls "success".

This is because gRPC calls are done inside ExecutorService, mostly via a ThreadPool. The Exceptions thrown inside a thread will "stay" inside that thread; when the Future.get() of that thread is called, the Exception will trigger an `ExecutionException` with the thread exception as cause.

In our code because we use CountDownLatch to help parallelize the gRPC call, we did not call the Future.get() to trigger the exception.

## Fix
Following https://stackoverflow.com/questions/19348248/waiting-on-a-list-of-future we rewrite to use the `ExecutorCompletionService<T>` to help out.

** We may need to look everywhere for CountDownLatch for potential similar problems**

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? No

## Release Notes

Fixed a bug where `GrpcBrokerRequestHandler` will ignore all gRPC exceptions.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java [hunk 1]
```diff
@@ -106,15 +108,17 @@ public BrokerResponseNative reduceOnStreamResponse(BrokerRequest brokerRequest,
...
-      ExecutionStatsAggregator aggregator) throws Exception {
+      ExecutionStatsAggregator aggregator) throws ExecutionException {
     int cnt = 0;
-    Future[] futures = new Future[serverResponseMap.size()];
-    CountDownLatch countDownLatch = new CountDownLatch(serverResponseMap.size());
-
+    Future<Void>[] futures = new Future[serverResponseMap.size()];
+    // based on ideas from on https://stackoverflow.com/questions/19348248/waiting-on-a-list-of-future
+    // mostly because we need to handle and transfer the exception in threads into caller thread.
+    ExecutorCompletionService<Void> countDownHelper = new ExecutorCompletionService<>(executorService);
```
Comment (dongxiaoman) [inline pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java:119]: Another way to fix is:
in the end of this method, call
```
for (Future future : futures) {
    future.get(reduceTimeoutMs, TimeUnit.MILLISECONS);
}
```
Somehow this has the potential to multiply the time wait by `futures.length()` times, and it breaks the purpose of parallel execution.

I remember in Scala I have written or found some utility class to help with this kind of `WaitForMultipleObjects` thing, not sure where to find it now.

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java [hunk 2]
```diff
@@ -127,23 +131,29 @@ private static void processIterativeServerResponse(StreamingReducer reducer, Exe
...
-        if (!future.isDone()) {
-          future.cancel(true);
+    for (int received = 0; received < futures.length; received++) {
+      try {
+        Future<Void> rpcCallHandle = countDownHelper.poll(reduceTimeOutMs, TimeUnit.MILLISECONDS);
+        if (rpcCallHandle == null) {
+          throw new TimeoutException("Failed to complete gRPC call within " + reduceTimeOutMs + "ms.");
+        }
+        // below will force the thread exception thrown into current thread
+        rpcCallHandle.get(reduceTimeOutMs, TimeUnit.MILLISECONDS);
```
Comment (dongxiaoman) [inline pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java:148]: @Jackie-Jiang By a quick code search, seems many places with `CountDownLatch` has similar pattern. They need to call `future.get()` to force the thread exception to throw, otherwise the calling thread (the one submitting the task) will consider the future success.
Example are https://github.com/apache/pinot/blob/b8fecb9d5a8fe524c42e29ce99ccf5012a2ebd49/pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GroupByDataTableReducer.java#L386

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java (hunk 1 above)
Comment (navina) [reply pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java:119]: wouldn't `CompletableFuture.allOf` satisfy your requirement here - https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#allOf-java.util.concurrent.CompletableFuture...- ?
Or is it too invasive change to replace `Future` with `CompletableFuture`?

Comment (dongxiaoman) [reply pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java]: Thought about this, at first thought gRPC calls returns a `Future<Void>` and it is painful to convert to `CompletableFuture`.
But I found https://stackoverflow.com/questions/23301598/transform-java-future-into-a-completablefuture , because we are creating Future by `ExecutorService.submit()` we can switch to use `CompletableFuture.supplyAsync()` to create CompletableFuture "natively".

Tried it out and it works. Now the code is shorter and cleaner.

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java [hunk 3]
```diff
@@ -106,15 +104,18 @@ public BrokerResponseNative reduceOnStreamResponse(BrokerRequest brokerRequest,
...
-    CountDownLatch countDownLatch = new CountDownLatch(serverResponseMap.size());
-
-    for (Map.Entry<ServerRoutingInstance, Iterator<Server.ServerResponse>> entry: serverResponseMap.entrySet()) {
-      futures[cnt++] = executorService.submit(() -> {
+    CompletableFuture<Void>[] futures = new CompletableFuture[serverResponseMap.size()];
+    // based on ideas from on https://stackoverflow.com/questions/19348248/waiting-on-a-list-of-future
+    // and https://stackoverflow.com/questions/23301598/transform-java-future-into-a-completablefuture
+    // Future created via ExecutorService.submit() can be created by CompletableFuture.supplyAsync()
+    for (Map.Entry<ServerRoutingInstance, Iterator<Server.ServerResponse>> entry : serverResponseMap.entrySet()) {
+      futures[cnt++] = CompletableFuture.supplyAsync(() -> {
```
Comment (richardstartin) [inline pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java:118]: This executes in a global pool which isn't the right behaviour. You can supply an executor to make it execute in the allotted thread pool.

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java [hunk 4]
```diff
@@ -106,15 +104,18 @@ public BrokerResponseNative reduceOnStreamResponse(BrokerRequest brokerRequest,
     return brokerResponseNative;
   }
 
-  private static void processIterativeServerResponse(StreamingReducer reducer, ExecutorService executorService,
```
Comment (richardstartin) [inline pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java:109]: This needs to be reverted so the work is done within `_reduceExecutorService`

Comment (dongxiaoman) [reply pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java:109]: Thanks. Didn't see that `supplyAsync( lambda, executor)` method. Fixed now

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java (hunk 3 above)
Comment (dongxiaoman) [reply pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java:119]: Fixed now. Just realized that there's such method.

**Review by richardstartin — APPROVED** [review]

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java [hunk 5]
```diff
@@ -127,23 +129,19 @@ private static void processIterativeServerResponse(StreamingReducer reducer, Exe
               aggregator.aggregate(entry.getKey(), dataTable);
             }
           }
+          return null;
```
Comment (walterddr) [inline pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java:132]: why return null here?

Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java:132]: +1, using `runAsync` instead would avoid this.

**Comment by dongxiaoman** [comment]
> StreamingReduceService.java

As I mentioned above, many places using `CountDownLatch` may have similar potential problem. We need to hunt them down and add unit tests one by one.

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java (hunk 5 above)
Comment (dongxiaoman) [reply pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java:132]: Yes good point. I think it was originally something with throwing Exceptions while `ExecutorService.submit` is called, and the Future<Void> asks for a return value (at least in my IDE it is red).
Anyway, changed now.

File: pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  @Test
+  public void testThreadExceptionTransfer() {
+    // simulate a thread exception in gRPC call and verify that the thread can transfer the exception
+    Iterator<Server.ServerResponse> mockedResponse = (Iterator<Server.ServerResponse>) mock(Iterator.class);
+    when(mockedResponse.hasNext()).thenReturn(true);
+    final String exceptionMessage = "Some exception";
+    final RuntimeException innerException = new RuntimeException(exceptionMessage);
+    when(mockedResponse.next()).thenThrow(innerException);
+    final ExecutorService threadPoolService = Executors.newFixedThreadPool(1);
+    final ServerRoutingInstance routingInstance =
```
Comment (richardstartin) [inline pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java:51]: It's not enforced but declaring variables final is against the project style, these are all effectively final so can be referenced in a nested scope anyway.

File: pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    // supposedly we can use TestNG's annotation like @Test(expectedExceptions = { IOException.class }) to verify
+    // here we hope to verify deeper to make sure the thrown exception is nested inside the exception
+    assertTrue(verifyException(() -> {
+          StreamingReduceService.processIterativeServerResponse(mock(StreamingReducer.class),
+              threadPoolService,
+              ImmutableMap.of(routingInstance, mockedResponse),
+              1000,
+              mock(BaseReduceService.ExecutionStatsAggregator.class));
+          return null;
+        }, (cause) -> cause.getMessage().contains(exceptionMessage))
```
Comment (richardstartin) [inline pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java:62]: unnecessary parentheses around lambda parameter

File: pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+              threadPoolService,
+              ImmutableMap.of(routingInstance, mockedResponse),
+              10,
+              mock(BaseReduceService.ExecutionStatsAggregator.class));
+          return null;
+        },
+        (cause) -> cause instanceof TimeoutException));
+  }
+
+  private static boolean verifyException(Callable<Void> verifyTarget, Function<Throwable, Boolean> verifyCause) {
```
Comment (richardstartin) [inline pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java:96]: Use Predicate<Throwable> instead

Comment (richardstartin) [reply pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java:96]: Also `Runnable` instead of `Callable<Void>`

File: pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+          throws Throwable {
+        Thread.sleep(1000);
+        return null;
+      }
+    });
+    final ExecutorService threadPoolService = Executors.newFixedThreadPool(1);
+    final ServerRoutingInstance routingInstance =
+        new ServerRoutingInstance("localhost", 9527, TableType.OFFLINE, false);
+    // supposedly we can use TestNG's annotation like @Test(expectedExceptions = { IOException.class }) to verify
+    // here we hope to verify deeper to make sure the thrown exception is nested inside the exception
```
Comment (richardstartin) [inline pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java:84]: It's hard to parse this sentence. I suggest changing it to something like "We can't use TestNG's expectedException because the exception we want to verify was thrown may be arbitrarily nested"

File: pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    }
+    try {
+      verifyTarget.call();
+    } catch (Exception ex) {
+      for (Throwable child = ex;
+          child != null && child.getCause() != child && !exceptionVerified;
+          child = child.getCause()) {
+        if (verifyCause.apply(child)) {
+          exceptionVerified = true;
+        }
```
Comment (richardstartin) [inline pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java:109]: `exceptionVerified = verifyCause.test(child)` since the loop terminates as soon as the predicate passes

**Review by richardstartin — APPROVED** [review]
I think the test code could be cleaned up a little if you feel like doing so

File: pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java (hunk 6 above)
Comment (dongxiaoman) [reply pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java]: Thanks. Did not know we prefer not using final on local variables. Fixed.

File: pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java (hunk 7 above)
Comment (dongxiaoman) [reply pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java]: Fixed

File: pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java (hunk 9 above)
Comment (dongxiaoman) [reply pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java]: Thanks. Rephrased

File: pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java (hunk 10 above)
Comment (dongxiaoman) [reply pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java]: Thanks. Shorter now

**Review by richardstartin — APPROVED** [review]
